src/RequestGroupMan.cc

```diff
@@ -474,11 +484,43 @@ createInitialCommand(const std::shared_ptr<RequestGroup>& requestGroup,
 void RequestGroupMan::fillRequestGroupFromReserver(DownloadEngine* e)
 {
   removeStoppedGroup(e);
-  if (static_cast<size_t>(maxSimultaneousDownloads_) <= numActive_) {
+
+  int maxSimultaneousDownloads=maxSimultaneousDownloads_;
+
+  if(optimizeSimultaneousDownloads_) {
+    // optimize method is requested, so gauge current speed
+    int currentSpeed=getNetStat().calculateDownloadSpeed();
+    if(currentSpeed>0) {
+      // apply the rule
+      int speed=currentSpeed;
+      if((maxOverallDownloadSpeedLimit_>0) && (speed>maxOverallDownloadSpeedLimit_)) {
+        speed=maxOverallDownloadSpeedLimit_;
+      }
+      maxSimultaneousDownloads=ceil(
+        optimizeSimultaneousDownloadsCoeffA_
+        +optimizeSimultaneousDownloadsCoeffB_ * log10(speed*8./1000000.)
+      );
+      if((maxSimultaneousDownloads_>0)&&(maxSimultaneousDownloads>maxSimultaneousDownloads_)) {
+        maxSimultaneousDownloads=maxSimultaneousDownloads_;
+      }
+      if(maxSimultaneousDownloads<1) {
+        maxSimultaneousDownloads=1;
+      }
+      A2_LOG_DEBUG
+        (fmt("Number of concurrent downloads adapted to speed (%d B/s) with max %d"
+             " (%lu currently active)",
+             speed, maxSimultaneousDownloads, numActive_));
```

Comment:
Perhaps, it would be nice to use `util::abbrevSize(speed).c_str()`, and replace `speed (%d B/s)` with `speed (%s B/s)` in case where speed gets faster.